[Security] Login link: Encrypt username

Nyholm opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.x for features
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | 
| License       | MIT
| Doc PR        | 

The current implementation of a "magic link" uses the following query parameters: "hash", "user" and "expires". This is fine but it leaves the username exposed to the world. It could arguable be fine to send an url like this to **foo@example.com**:

> https://app.com/magic?user=foo@example.com&expires=123&hash=xxxxxx

However, if the `UserInterface::getUsername()` does not return an email, but an other id, like a auth0 id. Im not sure it is a good idea to show publicly.

This PR uses sodium to do a symmetric encryption of the username. 

Potential improvements: 
- Make this opt-in or opt-out
- Move the logic to a new class
- Use sodium to encrypt the hash or the full query string

Nyholm committed:
[Security] Login link: Encrypt username

Nyholm committed:
cs fixes

Nyholm committed:
Removed mb_ prefix

nicolas-grekas commented:
Do we need to provide a layer that accepts the old links?

inmarelibero committed:
fixed potentially thrown \ErrorException and \SodiumException

inmarelibero commented:
hi @Nyholm I was reviewing your PR and found a couple of potencial bugs and I took the liberty to propose you a PR on your repo, you can find it at https://github.com/Nyholm/symfony/pull/10

for me anyway, +1 to use Sodium to encrypt the full query string, in order to have links like `login_check?code=[encrypted JSON of user, has, expiration, etc...]`

Nyholm commented:
Thank you. I agree with your changes. 

I'll do some more work on this PR to make sure we can do this change and still keep BC.

inmarelibero commented:
@Nyholm glad I helped in some way :) I'm looking into the tests, because it should have been caught I guess.

Feel free to ping me if you want me to do something

Nyholm commented:
Give me 30 minutes to push some new changes. I would be happy to hear your opinions about them.

Nyholm committed:
Refactor a bit

Nyholm committed:
CS

Nyholm committed:
cs

Nyholm committed:
Use more base64 encoding

Nyholm committed:
Added exceptions

inmarelibero commented on src/Symfony/Component/Security/Http/LoginLink/LoginLinkHandler.php:
```diff
@@ -71,7 +75,15 @@ public function createLoginLink(UserInterface $user): LoginLinkDetails
+            $username = $this->encryption->decrypt($usernameMessage);
+        } catch (MalformedCipherException $exception) {
+            // Keep compatibility with older versions of the code
+            $username = $usernameMessage;
```
@Nyholm what about explicitly indicate a deprecation?

because from "now" on, given https://github.com/symfony/symfony/pull/39275/files#diff-9a769cb79d1b50d804969e256e0443b7d98bc0e6bf421778ad3b22aa02849507R62, the `user` parameter will be always encoded and should never enter here

also, I was wondering if the solution of a `MalformedCipherException ` should or not be `$username = $usernameMessage`. In other words, a cipher could be malformed for many reasons, not only that `user` parameter is passed in clear

inmarelibero replied:
could we think about a `UsernameNotEncryptedException` ?

Nyholm replied:
I thought about this too. 
But I guess that people would generate a link, then upgrade and then user consumes the link. I just want to make sure the link works. If we didn't care about BC, I would always throw `InvalidLoginLinkException` here. Doing `$username = $usernameMessage` would also lead to a `InvalidLoginLinkException` if the user parameter wasnt encrypted. 

Sure, I can trigger a deprecation notice.

Nyholm committed:
Added comments

inmarelibero commented on src/Symfony/Component/Security/Core/Encryption/SymmetricEncryption.php:
```diff
@@ -0,0 +1,76 @@
+use Symfony\Component\Security\Core\Exception\MalformedCipherException;
+use Symfony\Component\Security\Core\Exception\UnsupportedAlgorithmException;
+
+class SymmetricEncryption
```
@Nyholm what about calling this `SymmetricEncrypter`?

inmarelibero replied:
I thought it would have been consistent with the names for example in https://github.com/Nyholm/symfony/tree/encrypt-username/src/Symfony/Component/Security/Core/Encoder, ending with `...er`

Nyholm replied:
That is true. You are correct. It just sounds strange.. 

I'll wait a bit an let it mature in my mind

OskarStark replied:
I like the ...er suffix 👍🏻 it reads more natural, it really do things an is not used todo things, in this case I would prefer SymmmetricEncryption 👍🏻

Nyholm committed:
Add deprecatio notive

Nyholm committed:
cs

jderusse commented on src/Symfony/Component/Security/Core/Encryption/SymmetricEncryption.php:
```diff
@@ -0,0 +1,85 @@
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Core\Encryption;
```
This class is only used in `security-http`. What's about marking it internal, and moving it in `Securty\Http`  namespace? In that way, we can also revert  `symfony/secrutiy-core: ^5.3` requirement.

Nyholm replied:
Yeah. But I really like this class. :)

I believe this is really useful outside the symfony namespace. Don’t you wake up every day and think about data you want to encrypt?

A follow up PR might even add asymmetric encryption. 

I didn’t want to create a new component and they don’t really belong in the http namespace.

chalasr replied:
Until we have an interface for this service (which could even push us to add `Sodium` to the classname/namespace), I'd suggest to make it internal.

OskarStark replied:
I agree with Tobias here, but let's add an interface which could be experimental in the first step. Afterwards we can change the signature of the LoginLink class to receive the interface instead of depending on the implementation.

Nyholm replied:
I want to insist. I've added an experimental tag and an interface.

inmarelibero commented:
@Nyholm it's taking shape, let me know if you want to delegate implementing tests, cheers

OskarStark commented on src/Symfony/Component/Security/Core/Encryption/SymmetricEncryption.php:
```diff
@@ -0,0 +1,85 @@
+    /**
+     * @var string application secret
+     */
+    private $secret;
```
I forget all the time, but can't we use property types for new features? 🧐

Nyholm replied:
Isn't that a 7.4 feature? We still need to support 7.2.

derrabus replied:
I'm looking forward to Symfony 6. 😈

OskarStark commented on src/Symfony/Component/Security/Core/Exception/DecryptionException.php:
```diff
@@ -0,0 +1,19 @@
+/**
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+class DecryptionException extends \Exception implements EncryptionExceptionInterface
```
It reads weird that Decrypt implements Encrypt 🙂

Maybe having separate exceptions is fine?

Nyholm replied:
Hm. Im not sure. But this is how I was reasoning: 

`EncryptionExceptionInterface` catches "all exceptions related to encryption". It is rare that you will do different actions if you are getting an exception when decrypting compared to encrypting. 

However, I agree with you that this is strange.. Maybe rename the `DecryptionException` to `EncryptionException` instead?

OskarStark commented on src/Symfony/Component/Security/Core/Exception/MalformedCipherException.php:
```diff
@@ -0,0 +1,23 @@
+{
+    public function __construct(\Throwable $previous = null)
+    {
+        parent::__construct('The message you provided does not look like a valid cipher text.', 0, $previous);
```
Maybe:

... is not a valid...

?

OskarStark commented on src/Symfony/Component/Security/Http/LoginLink/LoginLinkHandler.php:
```diff
@@ -33,8 +35,9 @@ final class LoginLinkHandler implements LoginLinkHandlerInterface
     private $secret;
     private $options;
     private $expiredStorage;
+    private $encryption;
```
encrypter?

Nyholm committed:
Added SymmetricEncryptionInterface

Nyholm committed:
Added some tests for the sodium encryption

Nyholm committed:
Added tests

Nyholm committed:
cs

Nyholm committed:
Updated service name

Nyholm committed:
Removed sodium

Nyholm committed:
Startign to use paragonie/halite

Nyholm committed:
Things are starting to work

Nyholm committed:
Fixes

Nyholm committed:
minor

Nyholm committed:
cs

Nyholm commented on src/Symfony/Component/Security/Http/LoginLink/LoginLinkHandler.php:
```diff
@@ -119,4 +125,32 @@ private function computeSignatureHash(UserInterface $user, int $expires): string
+
+    private function createKey(): EncryptionKey
+    {
+        // TODO make sure we dont use a static salt.
+        // TODO store salt some how between requests
+        $salt = "\x00\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f";
+
+        return KeyFactory::deriveEncryptionKey(new HiddenString($this->secret), $salt, KeyFactory::MODERATE);
```
I need help to figure this out. Using a static salt is not a good idea. 

The salt also needs to be kept secret. We must store the salt somehow.

sstok replied:
As a salt is merely used to prevent that a same value would always produce the same hash/key. Would it make sense to embed with the message? Encryption uses an Initialization Vector for a similar purpose, and that is part of the message as well.

Maybe you can ping the creator behind Halite for some advice?

Nyholm commented:
I removed all encryption stuff to rely on a third party package instead. 

I moved the encryption stuff to https://github.com/symfony/symfony/pull/39344

sstok commented:
> Do we need to provide a layer that accepts the old links?

I don't think that's a good idea, better to request a new link to than to support an old "insecure" link.

wouterj commented:
Hi @Nyholm! What is the status here? It seems like it has sort of stalled, although the changes seem to be complete. Is this ready to vote/merge?
